airflow/contrib/operators/adls_list_operator.py

```diff
@@ -46,7 +48,7 @@ class AzureDataLakeStorageListOperator(BaseOperator):
                 azure_data_lake_conn_id='azure_data_lake_default'
             )
     """
-    template_fields = ('path',)
+    template_fields = ('path',)  # type: Sequence[str]
```

Comment:
Do we need to do this on every operator? :(

Reply:
Seems like we need to set types explicitly for classes that we subclass elsewhere. Defining `template_fields` here gives it a type of `Tuple[str]`, which is incompatible with the type `Tuple[str, str]` in the subclass.

Reply:
What if we define the type on the base class as `List[str]` -- it doesn't need to explicitly be a tuple, just an iterable of strings. Does that help?

(It feels like for the type to be "right" since it is used in the base class we shouldn't re-define the type in the subclass)

Reply:
I think that would work, but I've been trying not to change the code more than I have to, so I didn't want to change tuples to lists in every operator. Do you think that would be a useful change?

Reply:
Oh I was thinking List was an abstract type (cos I call them arrays in my head). Maybe I mean `Iterable[str]` then? 

Not changing code sounds like an idea

Reply:
@jmcarp What about this comment? If we change this type to `Iterable[str]` can we avoid having to change code or add comment to every operator?

Reply:
We don't have to add type annotations for every operator--we have to add annotations on operators that are later subclassed. Changing the annotation to `Iterable[str]` doesn't change this. The issue is that a subclass defining `template_fields` as e.g. `("path", )` overrides its type from `Sequence[str]` to `Tuple[str]`; when the subclass of the subclass overrides the value to e.g. `("src_adls", "dest_gcs")`, its new type of `Tuple[str, str]` is incompatible with `Tuple[str]`. We could also set the types in `BaseOperator` to be `List[str]` like we talked about earlier, but that would mean changing all operators to use lists, which I think we agree we shouldn't do.

Here's a simplified example, using python3-style annotations for convenience:

```python
from typing import Iterable

class A:
    a: Iterable[str] = ("a", )

class B(A):
    a = ("a", "b")

class C(B):
    a = ("a", "b", "c")
```

This isn't valid, because the class variable gets typed as `Tuple[str, str]` in `B` but `Tuple[str, str, str]` in `C`:

```
error: Incompatible types in assignment (expression has type "Tuple[str, str, str]", base class "B" defined the type as "Tuple[str, str]")
```

If we annotate `B.a` with `Iterable[str]`, then mypy passes.

Altogether, the annotations we have so far are the simplest solution that I can think of without a larger refactor: we only have to annotate variables in a handful of classes (`BaseOperator` and other operators that are later subclassed), and we don't have to change tuples to lists for all operators.

Reply:
Whaaaaat. That strikes me as a bug in mypy - since there is no type annotation on B.a it should keep the one from A. Oh well.

Thanks for the explanation.

Still, let's change for `Iterable[Str]` where we do have to have an  annotation (over `Sequence`) - I reported/asked about this in https://github.com/python/mypy/issues/6511